pkg/event/registry.go

```diff
@@ -0,0 +1,163 @@
+package event
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/submariner-io/admiral/pkg/log"
+	k8sV1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+
+	submV1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+)
+
+type Registry struct {
+	eventHandlers []Handler
+	name          string
+	networkPlugin string
+}
+
+func NewRegistry(name, networkPlugin string) Registry {
+	return Registry{
+		name: name,
+		networkPlugin: networkPlugin,
+	}
+}
+
+// Adds a supported driver, prints a fatal error in the case of double registration
+func (er *Registry) AddHandler(eventHandler Handler) {
+	evNetworkPlugin := eventHandler.GetNetworkPlugin()
+
+	if evNetworkPlugin == AnyNetworkPlugin || evNetworkPlugin == er.networkPlugin {
+		er.eventHandlers = append(er.eventHandlers, eventHandler)
+		klog.Infof("Event handler %q loaded for the %q registry ",
+			eventHandler.GetName(), er.name)
+	} else {
+		klog.V(log.DEBUG).Infof("Event handler %q ignored for the %q registry ",
+			eventHandler.GetName(), er.name)
+	}
+}
+
+func (er *Registry) TransitionToNonGateway() error {
+	errors := map[string] error{}
+
+	for _, h := range er.eventHandlers {
+		errors[h.GetName()] = h.TransitionToNonGateway()
+	}
+
+	return flattenErrors(errors, "TransitionToNonGateway")
```

Comment:
We can make a generic function and use k8s `NewAggregate` instead of `flattenErrors`, ie

```
k8serrors "k8s.io/apimachinery/pkg/util/errors"
...

func (er *Registry) invokeHandlers(eventName string, invoke func(h Handler) error) error {
	var errs []error

	for _, h := range er.eventHandlers {
		err := invoke(h)
		if err != nil {
			errs = append(errs, errors.Wrapf(err, "%q returned error", h.GetName()))
		}
	}

	return errors.Wrapf(k8serrors.NewAggregate(errs), "%s failed", eventName)
}
```

Then this method becomes:

  ```
  return er.invokeHandlers("TransitionToNonGateway", func(h Handler) error {
		return h.TransitionToNonGateway()
	})
```

Reply:
Yes, that works only for the parameter-less, we'd need one invokeHandlers by type of function, unless there is some go-magic that would allow us to do a more dynamic thing... that would be perfect.

Reply:
I thought of this, but ... at least we save 1:3 of the code... so it's worth it, I'll update this.

I didn't know we had an error aggregate helper, that's cool

Reply:
Ahh, no with the lambda function we can handle the parameters, very clever :+1: :heart:

Reply:
Applied